cclib/method/ddec.py

```diff
@@ -271,59 +281,105 @@ def calculate(self, indices=None, fupdate=0.05):
+                # to second cycle
             ):
                 update_kappa = False
-                kappa = [0.0 for x in self.data.atomnos]
+                self._kappa = [0.0 for x in self.data.atomnos]
 
-            # TODO: EXIT LOOP HERE for steps = 7
             if not update_kappa:
+                # Increment steps
                 steps = steps + 1
+                # Calculate G_A and H_A to update w_A based on S4.3 in doi: 10.1039/c6ra04656h
+                self.reshape_G()
+                self.calculate_H()
+
             else:
                 # `steps` not incremented in this case
                 # First update kappa based on equation 93
-                kappa_new = kappa - self.N_A / self.u_A
-                kappa = [x if x > 0 else 0.0 for x in kappa_new]
-                # TODO: In both cases, calculate G_A and H_A to update w_A based on S4.3 in doi: 10.1039/c6ra04656h
+                kappa_new = self._kappa - self.N_A[cycle] / self.u_A[cycle]
+                self._kappa = [x if x > 0 else 0.0 for x in kappa_new]
+
+            # Update weights (w_A) using equation 96 in doi: 10.1039/c6ra04656h
+            for atomi in range(len(self._cond_density)):
+                self._cond_density[atomi] = math.exp(self._kappa[atomi]) * self._h[atomi]
+
+            # Update rho_cond for next iteration
+            self._update_rho_cond()
+
+        # * STEP 7 *
+        # In step 7, calculate w and u. Then, stop to calculate refcharges.
+        self.logger.info("Optimizing grid weights. (Step 7/7)")
+        self.N_A.append(self._calculate_w_and_u())
+
+        # Finally, store calculated DDEC6 charges in fragcharges
+        self.logger.info("Creating fragcharges: array[1]")
+        self.fragcharges = numpy.array(self.data.atomnos - self.N_A[-1], dtype=float)
 
     def calculate_refcharges(self):
         """ Calculate reference charges from proatom density and molecular density
             [STEP 1 and 2]
+            
+            Function returns calculated reference charges, localized charges, and stockholder
+            charges.
         """
         # Generator object to iterate over the grid
-        xshape, yshape, zshape = self.chgdensity.data.shape
-        natoms = self.data.natom
+        ngridx, ngridy, ngridz = self.chgdensity.data.shape
         indices = (
             (i, x, y, z)
-            for i in range(natoms)
-            for x in range(xshape)
-            for y in range(yshape)
-            for z in range(zshape)
+            for i in range(self.data.natom)
+            for x in range(ngridx)
+            for y in range(ngridy)
+            for z in range(ngridz)
         )
 
-        stockholder_w = numpy.zeros((natoms, xshape, yshape, zshape))
-        localized_w = numpy.zeros((natoms, xshape, yshape, zshape))
-        self.closest_r_index = numpy.zeros((natoms, xshape, yshape, zshape), dtype=int)
+        stockholder_w = numpy.zeros((self.data.natom, ngridx, ngridy, ngridz))
```

Comment:
Create a shape object and reuse it.

Reply:
I've added `grid_shape` for this. Thanks.